src/librustc_macros/src/query.rs

```diff
@@ -0,0 +1,122 @@
+use proc_macro::TokenStream;
+use syn::{Ident, ForeignItemFn, ReturnType};
+use syn::parse::{Result, Parse, ParseStream};
+
+struct List<T>(Vec<T>);
+
+impl<T: Parse> Parse for List<T> {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let mut list = Vec::new();
+        while !input.is_empty() {
+            list.push(input.parse()?);
+        }
+        Ok(List(list))
+    }
+}
+
+struct Group {
+    name: Ident,
+    queries: List<ForeignItemFn>,
+}
+
+impl Parse for Group {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let name: Ident = input.parse()?;
+        let content;
+        braced!(content in input);
+        Ok(Group {
+            name,
+            queries: content.parse()?,
+        })
+    }
+}
+
+fn camel_case(string: &str) -> String {
```

Comment:
I don't think we should do this. We already use the `snake_case` query name for `struct` / `enum` variant names, it's just dep nodes that are different because they started outside of the query system.

One (significant?) advantage of not changing the case of query names, at all, is greppability: `type_of` is `type_of` *everywhere* the query is mentioned, not `TypeOf` in some places.

cc @nikomatsakis

Reply:
That only really matters to incremental tests though. I'm not sure I actually want them to show up in search results =P

Reply:
I think we should either fully change it to keep `snake_case`, or keep the original behavior (so append `Item` to the name as to not change any test)

Reply:
I've changed the code to preserve the names.